lib/Dialect/Synth/SynthOps.cpp

```diff
@@ -67,12 +68,78 @@ llvm::APInt MajorityInverterOp::evaluate(ArrayRef<APInt> inputs) {
 OpFoldResult MajorityInverterOp::fold(FoldAdaptor adaptor) {
   // TODO: Implement maj(x, 1, 1) = 1, maj(x, 0, 0) = 0
 
+  // x x 1 -> x
+  // x 1 1 -> 1
+  // x ~x 1 -> 1
+  // x 0 0 -> 0
+  // x 1 1 -> 1
+
+  if (getNumOperands() != 3)
+    return {};
+  bool isOpConstant = true;
+  // for all constant inputs
   SmallVector<APInt, 3> inputValues;
+  size_t i = 0;
   for (auto input : adaptor.getInputs()) {
     auto attr = llvm::dyn_cast_or_null<IntegerAttr>(input);
     if (!attr)
-      return {};
-    inputValues.push_back(attr.getValue());
+      isOpConstant = false;
+    else {
+      auto value = isInverted(i) ? ~attr.getValue() : attr.getValue();
+      inputValues.push_back(value);
+    }
+    i++;
+  }
+  if (!isOpConstant) {
+    // x 0 0
+    // x 1 1
+    if (inputValues.size() == 2) {
+      if (inputValues[0] != inputValues[1])
+        return {};
```

Comment:
I think it's pessimistic. If two constants are same you can return constants, if they are complement you can return third variable.